### PR TITLE
Cat metadata

### PR DESCRIPTION
**Pull request by JimHofman**
Basic functionality and unit tests in place.

### CODE REVIEW

**Review by Anthony2k8900 — CHANGES_REQUESTED** [review]
*removed*

File: src/opera/pge/base_pge.py [hunk 1]
```diff
@@ -203,14 +206,40 @@ class and its inheritors, and as such, this class assumes access to the
     def __init__(self):
         self.name = None
         self.logger = None
+        self.runconfig_path = None
+        self.runconfig = RunConfig(self.runconfig_path)
```
Comment (collinss-jpl) [inline src/opera/pge/base_pge.py:210]: `PostProcessorMixin` should not have to define its own `runconfig` attribute since it `PgeExecutor` defines its own. Lines 209 and 210 should be removed. References to `self.runconfig` in `_create_catalog_metadata` should still work and reference the RunConfig created by `PgeExecutor`

File: src/opera/pge/base_pge.py [hunk 2]
```diff
@@ -203,14 +206,40 @@ class and its inheritors, and as such, this class assumes access to the
...
     def _run_sas_qa_executable(self):
         # TODO
         pass
 
     def _create_catalog_metadata(self):
-        # TODO
-        pass
+        """Returns the catalog metadata as a Python dictionary"""
+
+        self.cached_catalog_metadata = {
```
Comment (collinss-jpl) [inline src/opera/pge/base_pge.py:220]: Shouldn't this be the `self.catalog_metadata` you allocated in ``__init__``? Also, since this function currently returns a static dictionary, do we even need `self.catalog_metadata` to hold onto it?

File: src/opera/pge/base_pge.py [hunk 3]
```diff
@@ -203,14 +206,40 @@ class and its inheritors, and as such, this class assumes access to the
...
+            'PGE_Name': self.runconfig.pge_name,
+            'PGE_Version': "1.0.test",
```
Comment (collinss-jpl) [inline src/opera/pge/base_pge.py:222]: Would be good to get a ``#TODO`` here so we remember to revisit once we've integrated proper versioning

File: src/opera/pge/base_pge.py [hunk 4]
```diff
@@ -203,14 +206,40 @@ class and its inheritors, and as such, this class assumes access to the
...
+            'Ancillary_Files': self.runconfig.get_ancillary_filenames(),
+            'Production_DateTime': get_catalog_metadata_datetime_str(self.production_datetime)
+        }
+        return self.cached_catalog_metadata
+
+    def _write_catalog_metadata(self):
+        """Create, write and validate the catalog metadata json file"""
+
+        met_dict = self._create_catalog_metadata()
+        met_filename = "test_catalog_metadata.json"
```
Comment (collinss-jpl) [inline src/opera/pge/base_pge.py:234]: Another ``#TODO`` here would be good once we've moved onto integrating file-naming conventions

File: src/opera/schema/catalog_metadata_schema.json [hunk 5]
```diff
@@ -0,0 +1,63 @@
+{
```
Comment (collinss-jpl) [inline src/opera/schema/catalog_metadata_schema.json:1]: 👍 Looks great

File: src/opera/scripts/pge_main.py [hunk 6]
```diff
@@ -112,7 +112,6 @@ def load_run_config_file(logger, run_config_filename):
     # Log the yaml config file that is used.
     run_config = RunConfig(run_config_filename)
     msg = f'RunConfig yaml file: {run_config_filename}'
-    logger.info("pge_main", ErrorCode.RUN_CONFIG_FILENAME, msg)
```
Comment (collinss-jpl) [inline src/opera/scripts/pge_main.py:115]: Why was this removed?

File: src/opera/util/metfile.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+"""
+
+
+import json
+import jsonschema
+import os
+
+
+class MetFile:
+    """Class read and write .met files."""
```
Comment (collinss-jpl) [inline src/opera/util/metfile.py:35]: Change to "Class *used to* read and write *.json* catalog metadata files."

File: src/opera/util/metfile.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    def __init__(self, met_file_name, met_dict={}):
```
Comment (collinss-jpl) [inline src/opera/util/metfile.py:37]: It's bad practice in python to use a mutable object (in this case a dictionary) as a default argument for `met_dict`. Have `None` be the default value and if `None` is provided, initialize `met_dict` to an empty dictionary in the body of ``__init__``

File: src/opera/util/metfile.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        """
+        Constructor get filename and initialize dictionary
+
+        Parameters
+        ----------
+        met_file_name : str, required
+            Name of the file that will be written to or read.
+
```
Comment (collinss-jpl) [inline src/opera/util/metfile.py:45]: Description for `met_dict` is missing here

File: src/opera/util/metfile.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        """
+
+        self._met_file = met_file_name
+        self.met_dict = met_dict
+        self.combined_error_msg = None
+
+    def set_key_value(self, key_name, val):
```
Comment (collinss-jpl) [inline src/opera/util/metfile.py:52]: Since this method is basically an alias for `dict.__setitem__`, it might be neat to replace this method with an actual ``__setitem__`` so we could use `MetFile` objects like they were dictionaries themselves.

File: src/opera/util/metfile.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+            merged_met_dict = {}
+            with open(self._met_file) as json_file:
+                file_met_dict = json.load(json_file)
+                merged_met_dict = file_met_dict.copy()
+                merged_met_dict.update(self.met_dict)
+            with open(self._met_file, "w") as f:
+                json.dump(merged_met_dict, f, indent=2, sort_keys=True)
+        else:
+            with open(self._met_file, "w") as f:
+                json.dump(self.met_dict, f, indent=2, sort_keys=True)
```
Comment (collinss-jpl) [inline src/opera/util/metfile.py:68]: Can remove some redundancy here like so:

```python
merged_met_dict = {}

if os.path.exists(self._met_file):
    with open(self._met_file) as json_file:
        file_met_dict = json.load(json_file)
        merged_met_dict = file_met_dict.copy()

merged_met_dict.update(self.met_dict)
        
with open(self._met_file, "w") as f:
    json.dump(merged_met_dict, f, indent=2, sort_keys=True)
```

File: src/opera/util/metfile.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    def read_met_file(self):
+        """Reads, returns an existing met file."""
+        if os.path.exists(self._met_file):
+            # Open JSON file
+            with open(self._met_file) as json_file:
+                self.met_dict = json.load(json_file)
+            return self.met_dict
+
+    @staticmethod
+    def get_schema_file_path(schema_path='src/opera//schema/catalog_metadata_schema.json'):
```
Comment (collinss-jpl) [inline src/opera/util/metfile.py:79]: The schema filepath should be stored as a class constant for `MetFile`, and it should leverage `resource_filename` from `pkg_resources` to determine the path for us automatically:

```python
from pkg_resources import resource_filename
...
class Metfile:
    ...
    SCHEMA_PATH = resource_filename('opera', 'schema/catalog_metadata_schema.json')
```
This method could then just return `SCHEMA_PATH`

File: src/opera/util/metfile.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        validator = jsonschema.validators.validator_for(schema)(schema)
+        # record errors (if any)
+        errors = validator.iter_errors(json_data)
+        self.combined_error_msg = '\n'.join(error.message for error in errors)
+        # success if combined_error_text is an empty string
+        success = len(self.combined_error_msg) == 0
+        if success:
+            return True
+        else:
+            return False
```
Comment (collinss-jpl) [inline src/opera/util/metfile.py:116]: Conditional could be collapsed into just `return success`

File: src/opera/test/pge/test_metfile.py [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+    def testMetFile(self):
+        """
+        Tests instantiation of MetFile class
+        Methods tested in MetFile class:
+            set_key_value()
+            write_met_file()
+            read_met_file()
+
+        """
+        met_file = './testMetFile.met'
```
Comment (collinss-jpl) [inline src/opera/test/pge/test_metfile.py:91]: Use `resource_filename` to get the location of this file.

**Review by collinss-jpl — CHANGES_REQUESTED** [review]
Overall this branch is moving in the right direction, but there are a number of improvements I think could be made.

Also FYI, I rebased this branch to incorporate the addition of the build/test pipeline for pull requests in Jenkins. If you log into our Jenkins CI server you should now see pipelines run for this branch/PR. We can cover this more in our next tag-up. Great work so far, keep it up!

File: src/opera/scripts/pge_main.py (hunk 6 above)
Comment (JimHofman) [reply src/opera/scripts/pge_main.py:115]: Must have been an unintentional delete

File: src/opera/pge/base_pge.py (hunk 1 above)
Comment (JimHofman) [reply src/opera/pge/base_pge.py:210]: Done

File: src/opera/pge/base_pge.py (hunk 2 above)
Comment (JimHofman) [reply src/opera/pge/base_pge.py:220]: Now using a local dict.

Comment (JimHofman) [reply src/opera/pge/base_pge.py:220]: deleted self.catalog_metadata

File: src/opera/pge/base_pge.py (hunk 3 above)
Comment (JimHofman) [reply src/opera/pge/base_pge.py:222]: Done

File: src/opera/pge/base_pge.py (hunk 4 above)
Comment (JimHofman) [reply src/opera/pge/base_pge.py:234]: Done

File: src/opera/util/metfile.py (hunk 8 above)
Comment (JimHofman) [reply src/opera/util/metfile.py:37]: Done

File: src/opera/util/metfile.py (hunk 9 above)
Comment (JimHofman) [reply src/opera/util/metfile.py:45]: Done - I also updated the write_met_file() and read_met_file() docstrings.

File: src/opera/util/metfile.py (hunk 11 above)
Comment (JimHofman) [reply src/opera/util/metfile.py:68]: Nice - I like this one.  :)

File: src/opera/util/metfile.py (hunk 12 above)
Comment (JimHofman) [reply src/opera/util/metfile.py:79]: Done

File: src/opera/util/metfile.py (hunk 13 above)
Comment (JimHofman) [reply src/opera/util/metfile.py:116]: 'Doh!'

File: src/opera/util/metfile.py (hunk 12 above)
Comment (JimHofman) [reply src/opera/util/metfile.py:79]: Done

File: src/opera/test/pge/test_metfile.py (hunk 14 above)
Comment (JimHofman) [reply src/opera/test/pge/test_metfile.py:91]: Done

File: src/opera/util/metfile.py (hunk 10 above)
Comment (JimHofman) [reply src/opera/util/metfile.py:52]: Done

File: src/opera/test/pge/test_metfile.py [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+
+        """
+        cls.starting_dir = abspath(os.curdir)
+        cls.test_dir = resource_filename(__name__, "")
+        cls.data_dir = join(cls.test_dir, "data")
+
+        os.chdir(cls.test_dir)
+
+        cls.working_dir = tempfile.TemporaryDirectory(
+            prefix="test_time_", suffix='temp', dir=os.curdir)
```
Comment (collinss-jpl) [inline src/opera/test/pge/test_metfile.py:59]: Leftover reference to "test_time_" here

File: src/opera/test/pge/test_metfile.py [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+            lines = (mf.readlines())
+        # Test string to look for string in the dictionary
+        test_str = '  "test key": "test value",\n'
+        self.assertIn(test_str, lines)
+        # Test read_met_file()
+        met_dict = met_data.read_met_file()
+        self.assertEqual(met_data['test key'], 'test value')
+
+        # Read another line to test for an existing file
+        met_dict = "test key 2", "test value 2"
```
Comment (collinss-jpl) [inline src/opera/test/pge/test_metfile.py:113]: Not sure what this line is doing or is still needed? `met_dict` gets assigned again on line 121 before its used with this tuple (?)

File: src/opera/util/metfile.py [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+        self._met_file = met_file_name
+        self.met_dict = met_dict
+        self.combined_error_msg = None
+
+    def __setitem__(self, key_name, value):
+        print('in __setitem__')
+        self.met_dict[key_name] = value
+
+    def __getitem__(self, key_name):
+        print('in __getitem__')
```
Comment (collinss-jpl) [inline src/opera/util/metfile.py:66]: This one too

File: src/opera/pge/base_pge.py (hunk 2 above)
Comment (collinss-jpl) [reply src/opera/pge/base_pge.py]: I still see usage of `self.catalog_metadata` in this method

File: src/opera/util/metfile.py (hunk 7 above)
Comment (collinss-jpl) [reply src/opera/util/metfile.py]: This docstring still reads a bit awkwardly to me: "Class read and write catalog metadata .met files"

File: src/opera/test/pge/base_pge_test/scratch/testMetFile.met [hunk 18]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (collinss-jpl) [inline src/opera/test/pge/base_pge_test/scratch/testMetFile.met:1]: It looks like this file is only meant to be a test file that got checked in by accident. The `test_metfile.testMetFile` function needs to be updated to check for and remove this file if present at the end of the test. The check-and-delete should always occur regardless of whether the test passes or fails:


```python
met_file = MetFileTestCase.TEST_MET_FILE
try:
   <tests with assert statements>
finally:
    if os.path.exists(met_file):
        os.unlink(met_file)
```

As-is, `test_metfile.py` is failing in the CI system since this file can exist from a previous test run and fails when a new test run tries to overwrite due to permission issues.

File: src/opera/test/pge/test_metfile.py [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+            log = log.read()
+
+        # Verify the catalog metadata json file creation was logged.
+        self.assertIn("Successfully created catalog metadata json file.", log)
+
+        # Change the schema file
+        met = MetFile
+        # save the schema file
+        schema_file = met.get_schema_file_path()
+        shutil.copyfile(schema_file, 'save_file.json')
```
Comment (collinss-jpl) [inline src/opera/test/pge/test_metfile.py:164]: I think this function needs to check for and remove `save_file.json` before its finished. See my comment here: https://github.com/nasa/opera-sds-pge/pull/38/files#r790985366

**Review by collinss-jpl — CHANGES_REQUESTED** [review]
The first round of changes look great, but it also looks like a couple things fell through the cracks, and some new issues with the unit tests have been introduced. Getting real close though.

File: src/opera/test/pge/test_metfile.py (hunk 19 above)
Comment (collinss-jpl) [reply src/opera/test/pge/test_metfile.py:164]: Better yet, don't modify the schema, but create dummy metadata files to test for valid and invalid versions

File: src/opera/test/pge/test_metfile.py [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+        Instantiates pge objects
+        Verifies that the catalog metadata json file is created
+        Verifies that the schema is correct
+        Verifies that an Error is written to the log file is there is a schema error in the catalog metadata file.
+
+        """
+        # instantiate a pge object
+        runconfig_path = join(self.data_dir, 'test_base_pge_config.yaml')
+
+        pge = PgeExecutor(pge_name='BasePgeTest', runconfig_path=runconfig_path)
```
Comment (collinss-jpl) [inline src/opera/test/pge/test_metfile.py:135]: Rather than start from executing a full `PgeExecutor`, this test could start from creating local `MetFile` objects, both valid and invalid, and just call their validation methods on them.

File: src/opera/pge/base_pge.py (hunk 2 above)
Comment (JimHofman) [reply src/opera/pge/base_pge.py]: A local dictionary is now assigned and returned

File: src/opera/test/pge/test_metfile.py (hunk 15 above)
Comment (JimHofman) [reply src/opera/test/pge/test_metfile.py:59]: prefix="test_time_"  modified to prrefix="test_met_file_"

File: src/opera/util/metfile.py [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+        """
+
+        if met_dict is None:
+            met_dict = {}
+        self._met_file = met_file_name
+        self.met_dict = met_dict
+        self.combined_error_msg = None
+
+    def __setitem__(self, key_name, value):
+        print('in __setitem__')
```
[the remark this answers is not in the document]
Comment (JimHofman) [reply src/opera/util/metfile.py:62]: Done

File: src/opera/util/metfile.py (hunk 17 above)
Comment (JimHofman) [reply src/opera/util/metfile.py:66]: Done

File: src/opera/util/metfile.py [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+    def __getitem__(self, key_name):
+        print('in __getitem__')
+        return self.met_dict[key_name]
+
+    @classmethod
+    def get_schema_file_path(cls):
+        """Returns the path to schema file"""
+        return cls.SCHEMA_PATH
+
+    def write_met_file(self):
```
[the remark this answers is not in the document]
Comment (JimHofman) [reply src/opera/util/metfile.py:74]: Done

[the remark this answers is not in the document]
Comment (JimHofman) [reply src/opera/util/metfile.py:74]: done

File: src/opera/test/pge/test_metfile.py (hunk 16 above)
Comment (JimHofman) [reply src/opera/test/pge/test_metfile.py:113]: This whole section was redone. The simple test met file is generated and tested during the test and is deleted when the TemporaryDirectory goes away.

File: src/opera/test/pge/base_pge_test/scratch/testMetFile.met (hunk 18 above)
Comment (JimHofman) [reply src/opera/test/pge/base_pge_test/scratch/testMetFile.met:1]: This file is not being used anymore, and it was deleted from git.

File: src/opera/test/pge/test_metfile.py [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+            Schema file for the catalog_metadata_schema check
+        replace_values : str
+            Text string to be added or modify the schema file
+
+        """
+        for line in fileinput.input(schema_file, inplace=True):
+            for search_text in replace_values:
+                replace_text = replace_values[search_text]
+                line = line.replace(search_text, replace_text)
+            print(line, end='')
```
[the remark this answers is not in the document]
Comment (JimHofman) [reply src/opera/test/pge/test_metfile.py:214]: That was in the 'change_schema()' method, and that was deleted.

File: src/opera/test/pge/test_metfile.py (hunk 19 above)
Comment (JimHofman) [reply src/opera/test/pge/test_metfile.py:164]: re-wrote the test to test a modified dummy metadata file.

File: src/opera/test/pge/test_metfile.py (hunk 20 above)
Comment (JimHofman) [reply src/opera/test/pge/test_metfile.py:135]: Now localizing the test to just the validate_json_file() method.

File: src/opera/util/metfile.py (hunk 7 above)
Comment (JimHofman) [reply src/opera/util/metfile.py]: Done

**Review by collinss-jpl — APPROVED** [review]
Branch looks great, thanks again for resolving everything
